File: plugins/react-native/android/src/main/java/com/strata/reactnative/StrataModule.kt
package com.strata.reactnative

import android.app.ActivityManager
import android.content.Context
import android.content.pm.ActivityInfo
import android.content.res.Configuration
import android.os.Build
import android.os.PowerManager
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.view.InputDevice
import android.view.KeyEvent
import android.view.MotionEvent
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap


/**
 * Strata React Native module for Android.
 *
 * Provides device detection, input handling, and haptic feedback.
 */
class StrataModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private val buttons = linkedMapOf(
        "a" to false,
        "b" to false,
        "x" to false,
        "y" to false,
        "l1" to false,
        "r1" to false,
        "l2" to false,
        "r2" to false,
        "start" to false,
        "select" to false,
        "dpadUp" to false,
        "dpadDown" to false,
        "dpadLeft" to false,
        "dpadRight" to false
    )

    private var leftStickX = 0.0
    private var leftStickY = 0.0
    private var rightStickX = 0.0
    private var rightStickY = 0.0
    private var leftTrigger = 0.0
    private var rightTrigger = 0.0

    override fun getName(): String = "StrataModule"

    /**
     * Get device profile information.
     */
    @ReactMethod
    fun getDeviceProfile(promise: Promise) {
        try {
            val hasGamepad = hasGamepadConnected()
            val result = Arguments.createMap().apply {
                putString("platform", "android")
                putString("deviceType", getDeviceType())
                putString("inputMode", if (hasGamepad) "gamepad" else "touch")
                putString("orientation", getOrientation())
                putBoolean("hasTouch", reactApplicationContext.packageManager.hasSystemFeature("android.hardware.touchscreen"))
                putBoolean("hasGamepad", hasGamepad)
                putDouble("screenWidth", getScreenWidth())
                putDouble("screenHeight", getScreenHeight())
                putDouble("pixelRatio", getPixelRatio())
                putMap("safeAreaInsets", getSafeAreaInsetsMap())
                putString("performanceMode", getPerformanceModeInternal())
            }
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("DEVICE_PROFILE_ERROR", "Failed to get device profile: ${e.message}")
        }
    }

    private fun getPerformanceModeInternal(): String {
        var isLowPowerMode = false
        val powerManager = reactApplicationContext.getSystemService(Context.POWER_SERVICE) as? PowerManager
        if (powerManager != null) {
            isLowPowerMode = powerManager.isPowerSaveMode
        }

        val mi = ActivityManager.MemoryInfo()
        val activityManager = reactApplicationContext.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
        activityManager?.getMemoryInfo(mi)

        return if (isLowPowerMode || mi.totalMem < 2L * 1024 * 1024 * 1024) {
            "low"
        } else if (mi.totalMem < 4L * 1024 * 1024 * 1024) {
            "medium"
        } else {
            "high"
        }
    }

    @ReactMethod
    fun getPerformanceMode(promise: Promise) {
        val map = Arguments.createMap()

        var isLowPowerMode = false
        val powerManager = reactApplicationContext.getSystemService(Context.POWER_SERVICE) as? PowerManager
        if (powerManager != null) {
            isLowPowerMode = powerManager.isPowerSaveMode
        }

        val mi = ActivityManager.MemoryInfo()
        val activityManager = reactApplicationContext.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
        activityManager?.getMemoryInfo(mi)

        map.putString("mode", getPerformanceModeInternal())
        map.putBoolean("isLowPowerMode", isLowPowerMode)
        map.putDouble("totalMemory", mi.totalMem.toDouble())
        promise.resolve(map)
    }

    @ReactMethod
    fun setOrientation(orientation: String) {
        val currentActivity = currentActivity ?: return
        val orientationConstant = when (orientation) {
            "portrait" -> ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
            "landscape" -> ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
            else -> ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
        }
        currentActivity.requestedOrientation = orientationConstant
    }

    @ReactMethod
    @Synchronized
    fun getInputSnapshot(promise: Promise) {
        val snapshot = Arguments.createMap().apply {
            putDouble("timestamp", System.currentTimeMillis().toDouble())
            putMap("buttons", getButtonsMap())
            putMap("leftStick", getStickMap(leftStickX, leftStickY))
            putMap("rightStick", getStickMap(rightStickX, rightStickY))
            putMap("triggers", getTriggersMap())
            putArray("connectedGamepads", getConnectedGamepadIdsArray())
        }

        promise.resolve(snapshot)
    }

    @ReactMethod
    fun getSafeAreaInsets(promise: Promise) {
        promise.resolve(getSafeAreaInsetsMap())
    }

    /**
     * Trigger haptic feedback.
     */
    @ReactMethod
    fun triggerHaptics(intensity: String, promise: Promise) {
        try {
            val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                val vibratorManager = reactApplicationContext.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
                vibratorManager.defaultVibrator
            } else {
                @Suppress("DEPRECATION")
                reactApplicationContext.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            }

            if (!vibrator.hasVibrator()) {
                promise.resolve(null)
                return
            }

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val effect = when (intensity) {
                    "light" -> VibrationEffect.createOneShot(50, VibrationEffect.DEFAULT_AMPLITUDE)
                    "heavy" -> VibrationEffect.createOneShot(200, VibrationEffect.DEFAULT_AMPLITUDE)
                    else -> VibrationEffect.createOneShot(100, VibrationEffect.DEFAULT_AMPLITUDE) // medium
                }
                vibrator.vibrate(effect)
            } else {
                @Suppress("DEPRECATION")
                val duration = when (intensity) {
                    "light" -> 50L
                    "heavy" -> 200L
                    else -> 100L
                }
                vibrator.vibrate(duration)
            }

            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("HAPTICS_ERROR", "Failed to trigger haptics: ${e.message}")
        } catch (e: Error) {
            promise.reject("HAPTICS_ERROR", "Critical error triggering haptics: ${e.message}")
        }
    }

    private fun getDeviceType(): String {
        val configuration = reactApplicationContext.resources.configuration

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R &&
            reactApplicationContext.packageManager.hasSystemFeature("android.hardware.sensor.hinge_angle")) {
            return "foldable"
        }

        return when {
            (configuration.screenLayout and Configuration.SCREENLAYOUT_SIZE_MASK) >= Configuration.SCREENLAYOUT_SIZE_LARGE -> "tablet"
            else -> "mobile"
        }
    }

    private fun getScreenWidth(): Double {
        val metrics = reactApplicationContext.resources.displayMetrics
        return metrics.widthPixels.toDouble() / metrics.density
    }

    private fun getScreenHeight(): Double {
        val metrics = reactApplicationContext.resources.displayMetrics
        return metrics.heightPixels.toDouble() / metrics.density
    }

    private fun getPixelRatio(): Double {
        return reactApplicationContext.resources.displayMetrics.density.toDouble()
    }

    private fun getOrientation(): String {
        return if (reactApplicationContext.resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            "landscape"
        } else {
            "portrait"
        }
    }

    private fun getSafeAreaInsetsMap(): WritableMap {
        val insetsMap = Arguments.createMap().apply {
            putDouble("top", 0.0)
            putDouble("right", 0.0)
            putDouble("bottom", 0.0)
            putDouble("left", 0.0)
        }
        val density = reactApplicationContext.resources.displayMetrics.density
        val activity = currentActivity

        if (activity != null) {
            val windowInsets = ViewCompat.getRootWindowInsets(activity.window.decorView)
            if (windowInsets != null) {
                val cutoutInsets = windowInsets.getInsets(WindowInsetsCompat.Type.displayCutout())
                insetsMap.putDouble("top", cutoutInsets.top.toDouble() / density)
                insetsMap.putDouble("right", cutoutInsets.right.toDouble() / density)
                insetsMap.putDouble("bottom", cutoutInsets.bottom.toDouble() / density)
                insetsMap.putDouble("left", cutoutInsets.left.toDouble() / density)
            }
        }

        if (insetsMap.getDouble("top") == 0.0) {
            val resourceId = reactApplicationContext.resources.getIdentifier("status_bar_height", "dimen", "android")
            if (resourceId > 0) {
                insetsMap.putDouble(
                    "top",
                    reactApplicationContext.resources.getDimensionPixelSize(resourceId).toDouble() / density
                )
            }
        }

        return insetsMap
    }

    private fun hasGamepadConnected(): Boolean {
        return getConnectedGamepadIds().isNotEmpty()
    }

    private fun getConnectedGamepadIds(): List<Int> {
        return InputDevice.getDeviceIds().mapNotNull { deviceId ->
            val device = InputDevice.getDevice(deviceId)
            if (device != null && isGamepadDevice(device)) deviceId else null
        }
    }

    private fun isGamepadDevice(device: InputDevice): Boolean {
        return isGamepadSource(device.sources)
    }

    private fun isGamepadSource(source: Int): Boolean {
        return (source and InputDevice.SOURCE_GAMEPAD) == InputDevice.SOURCE_GAMEPAD ||
            (source and InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK
    }

    private fun getConnectedGamepadIdsArray() = Arguments.createArray().apply {
        getConnectedGamepadIds().forEach { pushInt(it) }
    }

    private fun getButtonsMap() = Arguments.createMap().apply {
        buttons.forEach { (button, pressed) -> putBoolean(button, pressed) }
    }

    private fun getStickMap(x: Double, y: Double) = Arguments.createMap().apply {
        putDouble("x", x)
        putDouble("y", y)
    }

    private fun getTriggersMap() = Arguments.createMap().apply {
        putDouble("left", leftTrigger)
        putDouble("right", rightTrigger)
    }

    @Synchronized
    fun handleKeyEvent(event: KeyEvent): Boolean {
        if (!isGamepadSource(event.source)) return false

        val button = when (event.keyCode) {
            KeyEvent.KEYCODE_BUTTON_A -> "a"
            KeyEvent.KEYCODE_BUTTON_B -> "b"
            KeyEvent.KEYCODE_BUTTON_X -> "x"
            KeyEvent.KEYCODE_BUTTON_Y -> "y"
            KeyEvent.KEYCODE_BUTTON_L1 -> "l1"
            KeyEvent.KEYCODE_BUTTON_R1 -> "r1"
            KeyEvent.KEYCODE_BUTTON_L2 -> "l2"
            KeyEvent.KEYCODE_BUTTON_R2 -> "r2"
            KeyEvent.KEYCODE_BUTTON_START -> "start"
            KeyEvent.KEYCODE_BUTTON_SELECT -> "select"
            KeyEvent.KEYCODE_DPAD_UP -> "dpadUp"
            KeyEvent.KEYCODE_DPAD_DOWN -> "dpadDown"
            KeyEvent.KEYCODE_DPAD_LEFT -> "dpadLeft"
            KeyEvent.KEYCODE_DPAD_RIGHT -> "dpadRight"
            else -> null
        }

        if (button == null) return false

        buttons[button] = event.action == KeyEvent.ACTION_DOWN
        return true
    }

    @Synchronized
    fun handleMotionEvent(event: MotionEvent): Boolean {
        if (!isGamepadSource(event.source) || event.action != MotionEvent.ACTION_MOVE) return false

        leftStickX = event.getAxisValue(MotionEvent.AXIS_X).toDouble()
        leftStickY = event.getAxisValue(MotionEvent.AXIS_Y).toDouble()
        rightStickX = event.getAxisValue(MotionEvent.AXIS_Z).toDouble()
        rightStickY = event.getAxisValue(MotionEvent.AXIS_RZ).toDouble()
        leftTrigger = maxOf(
            event.getAxisValue(MotionEvent.AXIS_BRAKE).toDouble(),
            event.getAxisValue(MotionEvent.AXIS_LTRIGGER).toDouble()
        )
        rightTrigger = maxOf(
            event.getAxisValue(MotionEvent.AXIS_GAS).toDouble(),
            event.getAxisValue(MotionEvent.AXIS_RTRIGGER).toDouble()
        )

        return true
    }
}

File: plugins/react-native/ios/StrataModule.swift
import Foundation
import GameController
import React
import UIKit

/**
 * Strata React Native module for iOS.
 *
 * Provides device detection, input handling, and haptic feedback.
 */
@objc(StrataModule)
class StrataModule: NSObject {

    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }

    /**
     * Get device profile information.
     */
    @objc
    func getDeviceProfile(_ resolve: @escaping RCTPromiseResolveBlock,
                          reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            let screen = UIScreen.main
            let hasGamepad = self.hasGamepadConnected()

            let result: [String: Any] = [
                "platform": "ios",
                "deviceType": self.getDeviceType(),
                "inputMode": hasGamepad ? "gamepad" : "touch",
                "orientation": self.getOrientation(),
                "hasTouch": true,
                "hasGamepad": hasGamepad,
                "screenWidth": screen.bounds.width,
                "screenHeight": screen.bounds.height,
                "pixelRatio": screen.scale,
                "safeAreaInsets": self.safeAreaInsetsMap(),
                "performanceMode": self.getPerformanceModeInternal()
            ]

            resolve(result)
        }
    }

    private func getPerformanceModeInternal() -> String {
        let isLowPowerMode = ProcessInfo.processInfo.isLowPowerModeEnabled
        if isLowPowerMode {
            return "low"
        }

        let physicalMemory = ProcessInfo.processInfo.physicalMemory
        if physicalMemory < 2 * 1024 * 1024 * 1024 {
            return "low"
        } else if physicalMemory < 4 * 1024 * 1024 * 1024 {
            return "medium"
        }

        return "high"
    }

    @objc
    func getPerformanceMode(_ resolve: @escaping RCTPromiseResolveBlock,
                           reject: @escaping RCTPromiseRejectBlock) {
        let isLowPowerMode = ProcessInfo.processInfo.isLowPowerModeEnabled

        resolve([
            "mode": self.getPerformanceModeInternal(),
            "isLowPowerMode": isLowPowerMode,
            "totalMemory": ProcessInfo.processInfo.physicalMemory
        ])
    }

    @objc
    func setOrientation(_ orientation: String) {
        DispatchQueue.main.async {
            var orientationValue: UIInterfaceOrientation = .unknown
            if orientation == "portrait" {
                orientationValue = .portrait
            } else if orientation == "landscape" {
                orientationValue = .landscapeLeft
            }

            if #available(iOS 16.0, *) {
                if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene {
                    let window = windowScene.windows.first
                    window?.rootViewController?.setNeedsUpdateOfSupportedInterfaceOrientations()
                }
            } else {
                UIDevice.current.setValue(orientationValue.rawValue, forKey: "orientation")
                UIViewController.attemptRotationToDeviceOrientation()
            }
        }
    }

    @objc
    func getInputSnapshot(_ resolve: @escaping RCTPromiseResolveBlock,
                          reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            resolve(self.createInputSnapshot())
        }
    }

    @objc
    func getSafeAreaInsets(_ resolve: @escaping RCTPromiseResolveBlock,
                           reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            resolve(self.safeAreaInsetsMap())
        }
    }

    /**
     * Trigger haptic feedback.
     */
    @objc
    func triggerHaptics(_ intensity: String,
                        resolve: @escaping RCTPromiseResolveBlock,
                        reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            let generator: UIImpactFeedbackGenerator

            switch intensity {
            case "heavy":
                generator = UIImpactFeedbackGenerator(style: .heavy)
            case "light":
                generator = UIImpactFeedbackGenerator(style: .light)
            default:
                generator = UIImpactFeedbackGenerator(style: .medium)
            }

            generator.prepare()
            generator.impactOccurred()

            resolve(nil)
        }
    }

    private func getDeviceType() -> String {
        if #available(iOS 14.0, *), UIDevice.current.userInterfaceIdiom == .mac {
            return "desktop"
        }

        switch UIDevice.current.userInterfaceIdiom {
        case .phone:
            return "mobile"
        case .pad:
            return "tablet"
        default:
            return "mobile"
        }
    }

    private func getOrientation() -> String {
        guard let windowScene = keyWindow()?.windowScene else {
            let bounds = UIScreen.main.bounds
            return bounds.width > bounds.height ? "landscape" : "portrait"
        }

        switch windowScene.interfaceOrientation {
        case .landscapeLeft, .landscapeRight:
            return "landscape"
        default:
            return "portrait"
        }
    }

    private func keyWindow() -> UIWindow? {
        if #available(iOS 13.0, *) {
            return UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .flatMap { $0.windows }
                .first { $0.isKeyWindow }
        }

        return UIApplication.shared.keyWindow
    }

    private func safeAreaInsetsMap() -> [String: CGFloat] {
        guard let window = keyWindow() else {
            return ["top": 0, "right": 0, "bottom": 0, "left": 0]
        }

        let insets = window.safeAreaInsets
        return [
            "top": insets.top,
            "right": insets.right,
            "bottom": insets.bottom,
            "left": insets.left
        ]
    }

    private func hasGamepadConnected() -> Bool {
        return !GCController.controllers().isEmpty
    }

    private func createInputSnapshot() -> [String: Any] {
        var buttons = defaultButtons()
        var leftStick: [String: Float] = ["x": 0, "y": 0]
        var rightStick: [String: Float] = ["x": 0, "y": 0]
        var triggers: [String: Float] = ["left": 0, "right": 0]

        if let controller = GCController.controllers().first {
            if let gamepad = controller.extendedGamepad {
                leftStick = [
                    "x": gamepad.leftThumbstick.xAxis.value,
                    "y": gamepad.leftThumbstick.yAxis.value
                ]
                rightStick = [
                    "x": gamepad.rightThumbstick.xAxis.value,
                    "y": gamepad.rightThumbstick.yAxis.value
                ]
                triggers = [
                    "left": gamepad.leftTrigger.value,
                    "right": gamepad.rightTrigger.value
                ]

                buttons["a"] = gamepad.buttonA.isPressed
                buttons["b"] = gamepad.buttonB.isPressed
                buttons["x"] = gamepad.buttonX.isPressed
                buttons["y"] = gamepad.buttonY.isPressed
                buttons["l1"] = gamepad.leftShoulder.isPressed
                buttons["r1"] = gamepad.rightShoulder.isPressed
                buttons["l2"] = gamepad.leftTrigger.isPressed
                buttons["r2"] = gamepad.rightTrigger.isPressed
                buttons["dpadUp"] = gamepad.dpad.up.isPressed
                buttons["dpadDown"] = gamepad.dpad.down.isPressed
                buttons["dpadLeft"] = gamepad.dpad.left.isPressed
                buttons["dpadRight"] = gamepad.dpad.right.isPressed

                if #available(iOS 12.1, *) {
                    buttons["leftStick"] = gamepad.leftThumbstickButton?.isPressed ?? false
                    buttons["rightStick"] = gamepad.rightThumbstickButton?.isPressed ?? false
                }

                if #available(iOS 13.0, *) {
                    buttons["start"] = gamepad.buttonMenu.isPressed
                    buttons["select"] = gamepad.buttonOptions?.isPressed ?? false
                }
            } else if let gamepad = controller.microGamepad {
                leftStick = [
                    "x": gamepad.dpad.xAxis.value,
                    "y": gamepad.dpad.yAxis.value
                ]
                buttons["a"] = gamepad.buttonA.isPressed
                buttons["x"] = gamepad.buttonX.isPressed
            }
        }

        let connectedGamepads = GCController.controllers().enumerated().map { index, controller in
            [
                "index": index,
                "id": controller.vendorName ?? "Game Controller \(index)"
            ] as [String: Any]
        }

        return [
            "timestamp": Date().timeIntervalSince1970 * 1000,
            "leftStick": leftStick,
            "rightStick": rightStick,
            "buttons": buttons,
            "triggers": triggers,
            "connectedGamepads": connectedGamepads
        ]
    }

    private func defaultButtons() -> [String: Bool] {
        return [
            "a": false,
            "b": false,
            "x": false,
            "y": false,
            "l1": false,
            "r1": false,
            "l2": false,
            "r2": false,
            "start": false,
            "select": false,
            "dpadUp": false,
            "dpadDown": false,
            "dpadLeft": false,
            "dpadRight": false
        ]
    }
}
